feat(shiny): add reactivity, testing, and agent-driving skills (#67)
posit-dev/skills: shiny/shiny-testing/SKILL.md

---

> name: shiny-testing

Reviewer: Let's rename this file & name value. Let's also add one for python

```suggestion
name: shiny-testing-r
```

description: Test Shiny apps in R across three layers — pure functions with testthat, reactive logic with shiny::testServer(), and end-to-end with shinytest2 (AppDriver, snapshots). Use when writing or fixing tests for a Shiny app or module, testing reactives without a browser, setting up shinytest2, or deciding what to test where. Triggers on testServer, shinytest2, AppDriver, snapshot tests, testing modules, and flaky Shiny tests.
metadata:
  author: Samuel Bharti (@samuelbharti)
  version: "1.0"
license: MIT
---

# Testing Shiny Apps

> Test in three layers, cheapest first. Most logic should be covered by the fast lower layers; reserve the slow browser layer for what genuinely needs rendering.

Reviewer: ```suggestion
Test in three layers, fastest first. Most logic should be covered by the fast lower layers; reserve the slow browser layer for what genuinely needs rendering.
```

`AppDriver` drives a real app in headless Chrome (via chromote), with idle-waiting built in.

```r
library(shinytest2)
test_that("app renders the filtered summary", {
  app <- AppDriver$new("path/to/app", name = "country-filter", seed = 1)
  app$set_inputs(country_filter = "Canada")  # auto-waits for the resulting change
  app$wait_for_idle()                         # extra settle for chained reactives
  app$expect_values()                         # JSON snapshot of inputs/outputs/exports
  app$expect_screenshot()                     # visual snapshot (review on first run)
})
```

Key `AppDriver` methods: `set_inputs()`, `wait_for_idle()`, `wait_for_value()`, `get_value(input=/output=/export=)`, `get_values()`, `get_logs()`, `expect_values()`, `expect_screenshot()`, `run_js()`.

- **Prefer `expect_values()` over `expect_screenshot()`** for data — value snapshots are stable; pixels flap across machines/fonts.
- Review snapshot changes with `testthat::snapshot_review()`.
- The driver waits on Shiny's own busy/idle signal, so you never write sleeps. (For driving a *live* app outside shinytest2, see **`shiny-for-agents`**.)

## Test behavior, not wiring

> Tests break when they assert *implementation* (internal IDs, intermediate values, exact HTML) instead of *behavior*. Assert what the user observes — the output value, the visible state — so refactors don't break green tests. Snapshot the result, not the plumbing.

Reviewer: ```suggestion
Tests break when they assert *implementation* (internal IDs, intermediate values, exact HTML) instead of *behavior*. Assert what the user observes — the output value, the visible state — so refactors don't break green tests. Snapshot app author values, not the plumbing.
```
